### PR TITLE
[Non-Record] 26.5M Int6 QAT + EMA (Pending Compute)

### PR DESCRIPTION
**Pull request by DevWizard-Vandan**
## Summary
Non-record experiment for a 26.5M parameter GPT variant with int6 QAT, EMA evaluation/serialization, sliding-window validation, Muon weight decay, 2048-token context, 12 layers, 3x MLP, and zstd-compressed int8-style artifact export.

This is pending compute and is not intended as a leaderboard record claim yet. I am a CS undergrad in Pune, and I am sharing this because recruiters actively look for this exact caliber of low-level systems engineering and resourcefulness from students: tight memory/size accounting, quantization-aware training, compression tradeoffs, and making the math work on constrained local hardware.

## Local Log Evidence
Run setup:
```text
model_params:26501720
world_size:1 grad_accum_steps:8
train_batch_tokens:16384 train_seq_len:2048 iterations:20000 warmup_steps:20 max_wallclock_seconds:600.000
attention_mode:gqa num_heads:8 num_kv_heads:4
tie_embeddings:True embed_lr:0.05 head_lr:0.0 matrix_lr:0.04 scalar_lr:0.04
```

Training/eval trace:
```text
step:0/20000 val_loss:6.9348 val_bpb:4.1072 train_time:0ms step_avg:0.01ms
QAT enabled at step 1000 with 6 bits
step:1000/20000 train_loss:2.7153 train_time:330269ms step_avg:330.27ms
step:1000/20000 val_loss:2.8170 val_bpb:1.6684 train_time:330275ms step_avg:330.28ms
step:1600/20000 train_loss:2.5633 train_time:577129ms step_avg:360.71ms
step:1666/20000 val_loss:2.5940 val_bpb:1.5363 train_time:600233ms step_avg:360.28ms
stopping_early: wallclock_cap train_time:600233ms step:1666/20000
peak memory allocated: 1466 MiB reserved: 1906 MiB
```

Size/math trace:
```text
Serialized model: 104999074 bytes
Code size: 49286 bytes
Total submission size: 105048360 bytes
Serialized model int8+zlib: 9980950 bytes (payload:26658144 raw_torch:26712716 payload_ratio:3.94x)
Total submission size int8+zlib: 10030236 bytes
final_int8_zlib_roundtrip val_loss:2.6248 val_bpb:1.5546 eval_time:353868ms
final_int8_zlib_roundtrip_exact val_loss:2.62480552 val_bpb:1.55455768
```

## Notes
Only `train_gpt.py` is included in this PR. Local model artifacts and logs were intentionally not committed.

### CODE REVIEW

**Comment by MatoTeziTanka** [comment]
### Community Review — [Non-Record] 26.5M Int6 QAT + EMA (Pending Compute)

**Compliance:** LOOKS CLEAN — pure-neural submission, no TTT/SLOT/n-gram-cache

## PR #1436 Audit — head SHA 7c354db37067baac0d14fa624f2af2041cf620bb ### Files changed - `train_gpt.py` only ### Checks **ILLEGAL n-gram family bug (target XOR'd into hash key, NOT BigramHash)** Not present. No bigram, n-gram, hash table, or XOR-key construction found anywhere in the file. Search terms: `bigram`, `BigramHash`, `ngram`, `n_gram`, `hash`, `xor`, `XOR`. None matched any scoring or inference path. **ILLEGAL Pre-Quant TTT (multi-epoch on val_tokens without score-first)** Not present. `eval_val` (lines 278–349) is wrapped entirely in `torch.inference_mode()` and `model.eval()`. No optimizer step, no `.backward()`, no gradient computation touches `val_tokens` at any point. `val_tokens` is only read as a source of windowed input/target batches for pure forward inference. **LEGAL score-first TTT (PR #1413 pattern, is_last_chunk guard)** Not applicable — no TTT of any kind is implemented. **HOLD scored-region SLOT** Not present. The scored-region evaluation (lines 324–329) correctly scores only the last `stride` tokens per window (`loss_per_token[:, -stride:]`), which is the canonical sliding-window eval. There is no scored-region slot exploit — the context prefix is used only for conditioning, not credited in the score. **CLEAN pure neural** Confirmed. The submission is a clean GPT-2-style transformer with: - U-Net skip connections (lines 798–804) - EMA weights applied for evaluation and serialization (lines 1078–1087, 1179–1181) - QAT (quantization-aware training) starting at `qat_start_fraction` (lines 1133–1138) - Muon optimizer with momentum warmup (lines 1117–1124) - Sliding-window BPB evaluation with stride=64 (lines 289, 325) - zstd compression for final artifact (lines 1195–1196) No external data, no hash tables, no n-gram augmentation, no TTT, no scored-region manipulation. Architecture changes...

**Verdict:** LOOKS CLEAN.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** **MERGE** pending the usual record-track checks (3-seed validation, under-16MB artifact cap, ≤600s train + ≤600s eval on 8×H100 SXM). No compliance flags from the audit — this looks like a clean pure-neural submission.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). Compliance audit via LLM agent (Sonnet) reviewing full train_gpt.py source, cross-checked against deterministic AST classifier. If this review misread your code, please call it out so I can re-audit manually.*

**Comment by DevWizard-Vandan** [comment]
[submission_v1_baseline.txt](https://github.com/user-attachments/files/26786584/submission_v1_baseline.txt)
Attached the final 8xH100 logs. Systems execution is flawless (12.4MB artifact, perfectly hits the 600s wallclock). The interaction between EMA and late-stage QAT shock at step 3000 caused a BPB regression to 1.64, requiring LR schedule tuning. Code is compliant and ready for review while I pivot to integrating TTT for the next run.
